File: backend/app/models/sale_item.py
# backend/app/models/sale_item.py
import uuid
from sqlalchemy import Column, Integer, Numeric, CheckConstraint, Text, text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy import ForeignKey
from sqlalchemy import TIMESTAMP
from sqlalchemy.sql import func

from app.db.base import Base


class SaleItem(Base):
    __tablename__ = "sale_items"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, server_default=text("gen_random_uuid()"))
    sale_id = Column(UUID(as_uuid=True), ForeignKey("sales.id"), nullable=False)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    quantity = Column(Integer, nullable=False)
    unit_price = Column(Numeric(14,2), nullable=False, server_default=text("0.00"))
    unit_cost = Column(Numeric(14,2), nullable=False, server_default=text("0.00"))
    discount = Column(Numeric, nullable=False, server_default=text("0.00"))
    tax_breakdown = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    line_total = Column(Numeric, nullable=False, server_default=text("0.00"))

    __table_args__ = (
        CheckConstraint('quantity > 0', name='sale_items_quantity_positive'),
    )

    # relationships
    sale = relationship("Sale", back_populates="items")
    product = relationship("Product", back_populates="sale_items")

File: j-backend/src/models/SaleItem.js
/**
 * SaleItem model - Line items in a sale.
 */
const { DataTypes } = require('sequelize');
const { sequelize } = require('../db');

const SaleItem = sequelize.define('SaleItem', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  sale_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'sales',
      key: 'id',
    },
  },
  product_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'products',
      key: 'id',
    },
  },
  quantity: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1,
    },
  },
  unit_price: {
    type: DataTypes.DECIMAL(14, 2),
    allowNull: false,
    defaultValue: 0.00,
  },
  unit_cost: {
    type: DataTypes.DECIMAL(14, 2),
    allowNull: false,
    defaultValue: 0.00,
  },
  discount: {
    type: DataTypes.DECIMAL(14, 2),
    allowNull: false,
    defaultValue: 0.00,
  },
  tax_breakdown: {
    type: DataTypes.JSONB,
    allowNull: false,
    defaultValue: {},
  },
  line_total: {
    type: DataTypes.DECIMAL(14, 2),
    allowNull: false,
    defaultValue: 0.00,
  },
}, {
  tableName: 'sale_items',
  timestamps: false,
});

module.exports = SaleItem;

